client/src/body/fabric-kernel.ts

```diff
@@ -56,37 +57,59 @@ function createOffsets(faceCountMax: number, fabricBytes: number): IOffsets {
     offsets.faceLocationsOffset = (
         offsets.faceNormalsOffset = (
             offsets.faceMidpointsOffset = (
-                offsets.vectorsOffset = 0
+                offsets.vectorsOffset = baseOffset
             ) + seedVectors * Float32Array.BYTES_PER_ELEMENT
         ) + faceVectorFloats * Float32Array.BYTES_PER_ELEMENT
     ) + faceJointFloats * Float32Array.BYTES_PER_ELEMENT
     return offsets
 }
 
-export class FabricKernel {
+export class FabricKernel implements IGotchiFactory {
     private instanceArray: IFabricInstanceExports[] = []
-    private offsets: IOffsets
+    private instanceUsed: boolean[] = []
     private arrayBuffer: ArrayBuffer
     private spotCenters: Float32Array
     private surface: Int8Array
 
-    constructor(exports: IFabricExports, dimensions: IFabricDimensions) {
+    constructor(private exports: IFabricExports, dimensions: IFabricDimensions) {
         const fabricBytes = exports.init(dimensions.jointCountMax, dimensions.intervalCountMax, dimensions.faceCountMax, dimensions.instanceMax)
         this.arrayBuffer = exports.memory.buffer
         this.spotCenters = new Float32Array(this.arrayBuffer, 0, SPOT_CENTERS_FLOATS)
         this.surface = new Int8Array(this.arrayBuffer, SPOT_CENTERS_SIZE, HEXALOT_BITS)
-        this.offsets = createOffsets(dimensions.faceCountMax, fabricBytes)
         const byteLength = this.arrayBuffer.byteLength
         if (byteLength === 0) {
-            throw new Error(`Zero byte length! ${this.offsets.fabricBytes}`)
+            throw new Error(`Zero byte length! ${fabricBytes}`)
         }
         for (let index = 0; index < dimensions.instanceMax; index++) {
-            this.instanceArray.push(new InstanceExports(this.arrayBuffer, this.offsets, exports, dimensions, index))
+            this.instanceArray.push(new InstanceExports(
+                this.arrayBuffer,
+                createOffsets(dimensions.faceCountMax, HEXALOT_SIZE + index * fabricBytes),
+                exports,
+                dimensions,
+                index,
+                toFree => this.instanceUsed[toFree] = false,
```

Comment:
Passing a callback to a constructor is a weird pattern.

Reply:
Why?